### PR TITLE
8313638: Add test for dump of resolved references

### PR DESCRIPTION
**Pull request by matias9927**
The change in [JDK-8306582](https://bugs.openjdk.org/browse/JDK-8306582) revealed that the state of the resolved references array is not checked in the CDS archive. This patch adds a test to ensure that the resolved references array is correct whether the application is archived or not.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313638](https://bugs.openjdk.org/browse/JDK-8313638): Add test for dump of resolved references (**Enhancement** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [99c18a91](https://git.openjdk.org/jdk/pull/15686/files/99c18a91ce60123d78c0b9d65d8242e251d68656)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [99c18a91](https://git.openjdk.org/jdk/pull/15686/files/99c18a91ce60123d78c0b9d65d8242e251d68656)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15686/head:pull/15686` \
`$ git checkout pull/15686`

Update a local copy of the PR: \
`$ git checkout pull/15686` \
`$ git pull https://git.openjdk.org/jdk.git pull/15686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15686`

View PR using the GUI difftool: \
`$ git pr show -t 15686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15686.diff">https://git.openjdk.org/jdk/pull/15686.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15686#issuecomment-1716088185)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/ResolvedReferencesWb.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import jdk.test.whitebox.WhiteBox;
+
+ public class ResolvedReferencesWb {
+    public static void main(String[] args) throws Exception {
+        WhiteBox wb = WhiteBox.getWhiteBox();
+
+        if (args.length < 2 && args[0].equals("--isArchived")) {
+            throw new RuntimeException("Test requires --isArchived flag");
+        }
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/ResolvedReferencesWb.java:32]: What happens if `args.length` is 0?
Also, the message for the `RuntimeExeption` could be clearer as follows:
   ` throw new RuntimeException("Test requires two args: --isArchived [true|false]");`
I think the "--isArchived" may not be needed, just passing in "true" or "false" should be sufficient.

**Review by calvinccheung — COMMENTED** [review]
Nice test addition. I have one comment. Looks good otherwise.

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/ResolvedReferencesWb.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                foundFoo |= (o.equals("fooString"));
+                foundBar |= (o.equals("barString"));
+                foundQux |= (o.equals("quxString"));
+            }
+        }
+
+        // CDS resolves all of the string literals used by the app and stores
+        // them inside the resolvedReferences array. The strings fooString and
+        // barString must have been found but not quxString unless ResolvedReferencesTestApp is archived.
+        if (isArchived) {
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/ResolvedReferencesWb.java:61]: I think the comment would be clear if you break it down in the two case, with something like this:

```
if (isArchived) {
    // CDS eagerly resolves all the string literals in the ConstantPool. At this point, all
    // three strings should be in the resolvedReferences array. 
} else {
    // If the class is not archived, the string literals in the ConstantPool are resolved
    // on-demand. At this point, ResolvedReferencesTestApp::<clinit> has been executed
    // so the two strings used there should be in the resolvedReferences array. 
    // ResolvedReferencesTestApp::qux() is not executed so "quxString"
    // should not yet be resolved.
}
```

**Review by iklam — COMMENTED** [review]
The test logic looks good to me. I think the comments can be improved.

**Review by iklam — APPROVED** [review]
LGTM. As we discussed off-line, the comments on line 26-30 are a little redundant so can be removed.

**Review by calvinccheung — APPROVED** [review]
Looks good. Thanks.

**Comment by matias9927** [comment]
Thank you for the reviews @calvinccheung and @iklam!
/integrate

**Pull request closed without merging**